sql/src/main/java/io/crate/execution/engine/window/WindowFunctionBatchIterator.java

```diff
@@ -0,0 +1,193 @@
+
+    public static BatchIterator<Row> of(BatchIterator<Row> source,
+                                        RowAccounting rowAccounting,
+                                        Comparator<Object[]> cmpPartitionBy,
+                                        Comparator<Object[]> cmpOrderBy,
+                                        int numCellsInSourceRow,
+                                        List<WindowFunction> windowFunctions,
+                                        List<? extends CollectExpression<Row, ?>> argsExpressions,
+                                        Input[]... args) {
+        // As optimization we use 1 list that acts both as inputs(source) and as outputs.
+        // The window function results are injected during the computation into spare cells that are eagerly created
+        Function<Row, Object[]> materialize = row -> materializeWithSpare(
+            rowAccounting.accountForAndMaybeBreak(row),
+            windowFunctions.size()
+        );
+        return CollectingBatchIterator.newInstance(
+            source,
+            src -> BatchIterators
+                .collect(src, Collectors.mapping(materialize, Collectors.toList()))
+                .thenApply(rows -> computeWindowFunctions(
+                    rows,
+                    cmpPartitionBy,
+                    cmpOrderBy,
+                    numCellsInSourceRow,
+                    windowFunctions,
+                    argsExpressions,
+                    args
+                ))
+                .thenApply(rows -> Lists.transform(rows, Buckets.arrayToSharedRow()::apply)),
+            source.involvesIO()
+        );
+    }
+
+    private static Object[] materializeWithSpare(Row row, int numWindowFunctions) {
+        Object[] cells = new Object[row.numColumns() + numWindowFunctions];
+        for (int i = 0; i < row.numColumns(); i++) {
+            cells[i] = row.get(i);
+        }
+        return cells;
+    }
+
+    static List<Object[]> computeWindowFunctions(List<Object[]> rows,
+                                                 @Nullable Comparator<Object[]> cmpPartitionBy,
+                                                 @Nullable Comparator<Object[]> cmpOrderBy,
+                                                 int numCellsInSourceRow,
+                                                 List<WindowFunction> windowFunctions,
+                                                 List<? extends CollectExpression<Row, ?>> argsExpressions,
+                                                 Input[]... args) {
+        if (rows.isEmpty()) {
+            return rows;
+        }
+        maybeSortRows(rows, cmpPartitionBy, cmpOrderBy);
+        int start = 0;
+        int end = rows.size();
+        int pStart = start;
+        int pEnd = findFirstNonPeer(rows, pStart, end, cmpPartitionBy);
+        var frame = new WindowFrameState(pStart, pEnd, rows);
+        boolean isTraceEnabled = LOGGER.isTraceEnabled();
```

Comment:
can be static final

Reply:
`isTraceEnabled` can change at runtime ( via `SET GLOBAL 'logger...`) . I do the assignment outside the inner loop as a trade-off between having the logging change visible between statement execution vs. *within* a statement being executed.